path_data_test.go

```diff
@@ -364,3 +498,447 @@ func TestVersionedKV_Reload_Policy(t *testing.T) {
+		},
+	}
+
+	req = &logical.Request{
+		Operation: logical.PatchOperation,
+		Path:      "data/foo",
+		Storage:   storage,
+		Data:      data,
+	}
+
+	resp, err = b.HandleRequest(context.Background(), req)
+
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("PatchOperation request failed - err:%s resp:%#v\n", err, resp)
+	}
+
+	req = &logical.Request{
+		Operation: logical.ReadOperation,
+		Path:      "data/foo",
+		Storage:   storage,
+	}
+
+	resp, err = b.HandleRequest(context.Background(), req)
+
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("Readperation request failed - err:%s resp:%#v\n", err, resp)
+	}
+
+	expectedData := map[string]interface{}{
+		"bar": "baz",
+		"abc": float64(123),
+		"quux": map[string]interface{}{
+			"def":  float64(456),
+			"quuz": []interface{}{"1", "2", "3", "4"},
+		},
+	}
+
+	if diff := deep.Equal(resp.Data["data"], expectedData); len(diff) > 0 {
+		t.Fatalf("secret data mismatch, diff: %#v\n", diff)
+	}
+}
+
+func TestVersionedKV_Patch_CurrentVersionDeleted(t *testing.T) {
+	b, storage := getBackend(t)
+
+	data := map[string]interface{}{
+		"data": map[string]interface{}{
+			"bar": "baz",
+		},
+	}
+
+	req := &logical.Request{
+		Operation: logical.CreateOperation,
+		Path:      "data/foo",
+		Storage:   storage,
+		Data:      data,
+	}
+
+	resp, err := b.HandleRequest(context.Background(), req)
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("CreateOperation request failed - err:%s resp:%#v\n", err, resp)
+	}
+
+	req = &logical.Request{
+		Operation: logical.DeleteOperation,
+		Path:      "data/foo",
+		Storage:   storage,
+		Data:      data,
+	}
+
+	resp, err = b.HandleRequest(context.Background(), req)
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("DeleteOperation request failed - err:%s resp:%#v\n", err, resp)
+	}
+
+	req = &logical.Request{
+		Operation: logical.ReadOperation,
+		Path:      "data/foo",
+		Storage:   storage,
+	}
+
+	resp, err = b.HandleRequest(context.Background(), req)
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("ReadOperation request failed - err:%s resp:%#v\n", err, resp)
+	}
+
+	// Use of logical.RespondWithStatusCode in handler will
+	// serialize the JSON response body as a string
+	respBody := map[string]interface{}{}
+
+	if rawRespBody, ok := resp.Data[logical.HTTPRawBody]; ok {
+		err = json.Unmarshal([]byte(rawRespBody.(string)), &respBody)
+	}
+
+	respDataRaw, ok := respBody["data"]
+	if !ok {
+		t.Fatalf("No data provided in response, resp: %#v\n", resp)
+	}
+
+	respData := respDataRaw.(map[string]interface{})
+
+	respMetadataRaw, ok := respData["metadata"]
+	if !ok {
+		t.Fatalf("No metadata provided in response, resp: %#v\n", resp)
+	}
+
+	respMetadata := respMetadataRaw.(map[string]interface{})
+
+	if respMetadata["deletion_time"] == "" {
+		t.Fatalf("Expected deletion_time to be set, resp:%#v\n", resp)
+	}
+
+	data["quux"] = "quuz"
+
+	req = &logical.Request{
```

Comment:
It isn't required for patch operations unless the mount or secret config has `cas_required` set to true. You calling this out has made me realize that we do not have any happy path tests that use the `cas` option. I will account for that.

Reply:
It turns out `TestVersionedKV_Patch_Success` already uses `cas`. My thinking is that it will be good to have some tests that specify a `cas` option and some that do not. What do you think?

Reply:
That'd be great!